cloud/scope/identity.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/pkg/errors"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	aadpodv1 "github.com/Azure/aad-pod-identity/pkg/apis/aadpodidentity/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AzureCredentialsProvider provides
+type AzureCredentialsProvider struct {
+	Client       client.Client
+	AzureCluster *infrav1.AzureCluster
+	Identity     *aadpodv1.AzureIdentity
+}
+
+// NewAzureCredentialsProvider creates a new AzureCredentialsProvider from the supplied inputs.
+func NewAzureCredentialsProvider(ctx context.Context, kubeClient client.Client, azureCluster *infrav1.AzureCluster, identityName, namespace string) (*AzureCredentialsProvider, error) {
+	if identityName == "" {
+		return nil, errors.New("failed to generate new AzureCredentialsProvider from empty identityName")
+	}
+
+	azureIdentity := &aadpodv1.AzureIdentity{}
+	err := kubeClient.Get(ctx, client.ObjectKey{Name: identityName, Namespace: namespace}, azureIdentity)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get AzureIdentity")
+	}
+	if azureIdentity.Spec.Type != aadpodv1.ServicePrincipal {
+		return nil, errors.New("AzureIdentity is not of type Service Principal")
+	}
+
+	return &AzureCredentialsProvider{
+		Client:       kubeClient,
+		AzureCluster: azureCluster,
+		Identity:     azureIdentity,
+	}, nil
+}
+
+// GetAuthorizer returnes Azure authorizer based on the provided azure identity
+func (p *AzureCredentialsProvider) GetAuthorizer(ctx context.Context, resourceManagerEndpoint string) (autorest.Authorizer, error) {
+	copiedIdentity := &aadpodv1.AzureIdentity{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AzureIdentity",
+			APIVersion: "aadpodidentity.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      p.Identity.Name,
```

Comment:
How do we handle `AzureIdentity` with same name in multi namespaces? When run in namespaced mode, user could configure `azid1` in `ns1` and `azid1` in `ns2` (both with different `clientIDs`).

cluster1 referring to `ns1/azid1` and cluster referring to `ns2/azid` will result in conflict while copied in controller namespace? Should we add the namespace to the copied `AzureIdentity` name?

Reply:
good idea, will add the namespace to the name